### PR TITLE
Enhancement/async property

### LINKED ISSUE #36: Async behavior
I have studied the led example. Although is uses a async webserver you still need to "POLL" propertys to detect changes. 
Should it not be better to provide possibilities for a callback when a property is changed ?

### PR DESCRIPTION
**Pull request by lgLindstrom**
I have added, what I call Asynch Property functionallity. Not being able to test anything else than the ESPWebThingAdapter.
Also added a example.

Fixes #36

### CODE REVIEW

File: ESPWebThingAdapter.h [hunk 1]
```diff
@@ -28,7 +28,7 @@
 
 class WebThingAdapter {
 public:
-  WebThingAdapter(String _name, IPAddress _ip): name(_name), server(80), ip(_ip.toString()) {
+  WebThingAdapter(String _name, IPAddress _ip): name(_name) , ip(_ip.toString()) , server(80) {
```
Comment (hobinjk) [inline ESPWebThingAdapter.h]: nit: remove space before comma after name(_name)

File: ESPWebThingAdapter.h [hunk 2]
```diff
@@ -242,16 +242,19 @@ class WebThingAdapter {
...
       case NUMBER: {
         ThingPropertyValue value;
         value.number = newValue.as<double>();
         property->setValue(value);
+        property->changed(value);
         break;
       }
       case STRING:
         *property->getValue().string = newValue.as<String>();
+        property->changed( property->getValue() );
```
Comment (hobinjk) [inline ESPWebThingAdapter.h]: nit: don't pad function arguments with spaces

File: EthernetWebThingAdapter.h [hunk 3]
```diff
@@ -392,16 +392,19 @@ class WebThingAdapter {
...
     case NUMBER: {
       ThingPropertyValue value;
       value.number = newValue.as<double>();
       property->setValue(value);
+      property->changed(value);
       break;
     }
     case STRING:
       *property->getValue().string = newValue.as<String>();
+      property->changed( property->getValue() );
```
Comment (hobinjk) [inline EthernetWebThingAdapter.h]: nit: don't pad function arguments with spaces

File: Thing.h [hunk 4]
```diff
@@ -32,11 +32,12 @@ class ThingProperty {
   String atType;
   ThingProperty* next = nullptr;
 
-  ThingProperty(const char* id_, const char* description_, ThingPropertyType type_, const char* atType_):
+ ThingProperty(const char* id_, const char* description_, ThingPropertyType type_, const char* atType_ , void (*callBack_) (ThingPropertyValue) = nullptr):
```
Comment (hobinjk) [inline Thing.h]: This is the wrong indentation level

File: Thing.h [hunk 5]
```diff
@@ -47,8 +48,16 @@ class ThingProperty {
     return this->value;
   }
 
+  void changed(ThingPropertyValue newValue) {
+    if ( callBack != nullptr ) {
```
Comment (hobinjk) [inline Thing.h]: nit: Don't pad parentheses with spaces

File: WiFi101WebThingAdapter.h [hunk 6]
```diff
@@ -365,16 +365,19 @@ class WebThingAdapter {
...
     case NUMBER: {
       ThingPropertyValue value;
       value.number = newValue.as<double>();
       property->setValue(value);
+      property->changed(value);
       break;
     }
     case STRING:
       *property->getValue().string = newValue.as<String>();
+      property->changed( property->getValue() );
```
Comment (hobinjk) [inline WiFi101WebThingAdapter.h]: nit: don't pad function arguments with spaces

File: examples/PlatformIO/AsyncProperty/.vscode/extensions.json [hunk 7]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (hobinjk) [inline examples/PlatformIO/AsyncProperty/.vscode/extensions.json]: Delete this file

File: examples/PlatformIO/AsyncProperty/.vscode/settings.json [hunk 8]
```diff
@@ -0,0 +1,6 @@
+{
```
Comment (hobinjk) [inline examples/PlatformIO/AsyncProperty/.vscode/settings.json]: Delete this file

File: examples/PlatformIO/AsyncProperty/asyncProperty.code-workspace [hunk 9]
```diff
@@ -0,0 +1,13 @@
+{
```
Comment (hobinjk) [inline examples/PlatformIO/AsyncProperty/asyncProperty.code-workspace]: Delete this file

File: examples/PlatformIO/AsyncProperty/include/README [hunk 10]
```diff
@@ -0,0 +1,39 @@
+
```
Comment (hobinjk) [inline examples/PlatformIO/AsyncProperty/include/README]: Delete this file

File: examples/PlatformIO/AsyncProperty/test/README [hunk 11]
```diff
@@ -0,0 +1,11 @@
+
```
Comment (hobinjk) [inline examples/PlatformIO/AsyncProperty/test/README]: Delete this file

File: examples/PlatformIO/AsyncProperty/src/main.cpp [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+
+  textDisplay.addProperty(&text);
+  textDisplay.addProperty(&onOff);
+  textDisplay.addProperty(&number);
+  adapter->addDevice(&textDisplay);
+  adapter->begin();
+}
+
+void loop() {
+  delay(500);
```
Comment (hobinjk) [inline examples/PlatformIO/AsyncProperty/src/main.cpp]: I believe that this needs a call to `adapter->update();`

File: examples/PlatformIO/AsyncProperty/src/main.cpp [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+  Serial.print("Connecting : ");
+  WiFi.begin(ssid, password);
+  
+
+  // Wait for connection
+  while (WiFi.status() != WL_CONNECTED) {
+
+    // more than 5 sec wait, try to login again.
+    if (loginCounter == 10) 
+    {
```
Comment (hobinjk) [inline examples/PlatformIO/AsyncProperty/src/main.cpp]: The brace style should be consistent with the rest of the code, which is
```
if (loginCounter == 10) {
} else {
}
```

File: examples/PlatformIO/AsyncProperty/platformio.ini [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+[env:esp32dev]
+platform = espressif32
+board = esp32dev
+framework = arduino
+lib_deps =
+    
+    ${global.lib_deps}
+    ESP Async WebServer
+lib_extra_dirs =
+    ../../../..
```
Comment (hobinjk) [inline examples/PlatformIO/AsyncProperty/platformio.ini]: I think this line might be causing the automated test failures

**Comment by lgLindstrom** [comment]
Okay,, I will correct.

question, is there a lint tool/setup to use ?

/mvh
/lg
_____________________________________________________
Lars-Göran Lindström
Consius AB
+46 70 190 23 30
lg.lindstrom@consius.se


Den mån 22 apr. 2019 kl 15:35 skrev James Hobin <notifications@github.com>:

**Comment by hobinjk** [comment]
The best style guide I could find is https://developer.mozilla.org/en-US/docs/Mozilla/Developer_guide/Coding_Style. I was hoping to track down an automated tool for this but was sidetracked by work on the main gateway.

**Comment by mrstegeman** [comment]
@lgLindstrom Are you still interested in this PR? If so, would you mind rebasing and cleaning things up so that it can pass the CI tests?

**Pull request merged**